lib/worker/exec.go

```diff
@@ -0,0 +1,129 @@
+package worker
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os/exec"
+	"strings"
+	"time"
+)
+
+// Exec represents something to be executed in the Worker
+// TODO: I might just remove this interface in the next
+// PR. I introduced it thinking it could be useful for
+// the future cgroup implementation but now am not so
+// sure.
+type Exec interface {
+	// Command is the command for the exec.
+	Command() string
+	// Start starts the exec.
+	Start() error
+	// Stop stops the exec.
+	Stop() error
+	// PID returns the pid of the process once it has been started.
+	// Otherwise, and invalid pid 0 is returned.
+	PID() int
+	// Status returns the ProcessStatus struct corresponding
+	// to the started exec.
+	Status() (*ProcessStatus, error)
+}
+
+// process is a process executed or to be executed by the worker
+// TODO: Find a better name for this as the word "process" implies
+// that this it is already running.
+type process struct {
+	command    string
+	cmd        *exec.Cmd
+	startedBy  string
+	startedAt  time.Time
+	finishedAt time.Time // TODO: implement finished at
+}
+
+func newProcess(command string, cmd *exec.Cmd, startedBy string) *process {
+	return &process{
+		command:   command,
+		cmd:       cmd,
+		startedBy: startedBy,
+	}
+}
+
+// ProcessStatus represents a snapshot of a process status
+type ProcessStatus struct {
+	PID        int
+	StartedBy  string
+	State      string
+	StartedAt  time.Time
+	FinishedAt time.Time // TODO: implement finished at
+}
+
+func (p *process) Start() error {
+	p.startedAt = time.Now()
+	err := p.cmd.Start()
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (p *process) Stop() error {
+	if p.cmd.Process == nil {
```

Comment:
Is this a race condition? Can `p.cmd.Process` ever be `nil` again after being set?

Reply:
It cannot be nil again, this is set by the `cmd.Start()` method when the process is started. However, it would be a bit confusing (and potentially problematic) to have Start and Stop overlapping so I will make these methods mutually exclusive.

Reply:
With my latest change (making Start and Stop mutually exclusive), it should be more safe to run Stop concurrently.